### PR TITLE
Second Refactor - science

### LINKED ISSUE #41: Refactor main parameterisation routine to deal in tendencies
As discussed at the end of #33 interfacing with CAM requires the parameterisation to output tendencies rather than updating variables.
This is at a complexity level below what is currently used for SAM.
Therefore we should refactor so that tendencies are returned, and any updating of variables (q, t) are done in the model interface.

Further detail from a discussion is as follows:

* `q` is total water, and `qn` is cloud water.  
    - `q` is total water content (vapour plus `qn`), `qn` is non-vapour water content.
    - We can probably comment out `qn` from main parameterisation as it comes from SAM setup - in larger models it will be immediately precipitated out during a larger timestep.
    - In fact, just delete `qn`!!
* Variables labelled 'tendencies' are (generally) $\Delta\psi$?
    - It would be nicer if we can refactor for tendencies to actually be tendencies rather than changes
    - The checks still operate on the actual value, however, so be aware of this - especially for potential subcycling!
    - SAM equations are of the form:  
    $\text{tendency}\psi = \text{d} \psi / \text{d}t = - 1 / \rho * \text{d} (\rho * u * \psi) / \text{d}x = - 1 / \rho * \text{d}(\text{flux}\psi) / \text{d}x$
      So the `irhoadzdz` term comes from rearranging to get:
      $\text{d}\psi = - \text{irhoadzdz} * \text{d}(\text{flux} \psi)$
    - CAM operates at the level of tendencies.
    - careful about removing dt as there is a dtn in irhodzdz
* t is `energy / cp` beware of this, especially in the SAM-CAM conversion

### PR DESCRIPTION
**Pull request by jatkinson1000**
Second refactor of code from a scientific standpoint rather than the structural standpoint of #33 

Tasks (discussed in #41 ):
- [X] Remove `qn`
- [X] Rename any mislabelled `tendency_` variables to `delta_`
- [x] Decide how to return (`q` and `t` or `delta` and `tendency`) based on requirements of SAM and CAM  
Does CAM provide something appropriate for `q` that we can use to enforce physics conditions?

closes #41

### CODE REVIEW

**Comment by jatkinson1000** [comment]
@paogorman I have quick question about the code here.
I have refactored so that the code returns tendencies of variables for use in CAM, rather than updating `t` and `q` internally.
However, the variables `t` and `q` are required to be updated within the physical layer in order to enforce some of the physical constraints (non-negative water content etc).
Are you happy with me saying that we will still use the '`t`' and '`q`' that have been `derived`/converted from CAM variables for this purpose, updating as we enforce various constraints, but these will then be discarded at the end of the subroutine with tendencies returned?

**Comment by paogorman** [comment]
Yes. I think we will need the updates in t and q internally but we can just forget these between calls to the parameterization.

**Comment by jatkinson1000** [comment]
@paogorman 
As part of these changes there are 2 approaches that could be taken:
- a) Pass the entire domain to the parameterisation and get out deltas for all columns.
- b) Loop over the columns calling the parameterisation for each one.

Is there a preferable option?
a) is likely more computationally efficient, but I seem to recall that operating on a 'per column basis' might be an advantage in CAM? Perhaps something @judithberner can help clarify later today?

**Comment by paogorman** [comment]
I recall the issue is that we ideally want the parameterization to work for both SAM and CAM going forward. SAM works based on (lon,lat,vertical level) whereas CAM has (pcol,level) where pcol is the number of columns. A simple option is to just do one column per call to the parameterization but we will presumably sacrifice computational efficiency. Alternatively we could make pcol the number of grid points in latitude when using the parameterization for SAM and have an outer loop for longitude (or the other way around if it is faster). Something to discuss in the meeting.

**Comment by jatkinson1000** [comment]
Haha, yes, I realised shortly before getting this reply that that was an option! 😅 Thanks!

**Comment by jatkinson1000** [comment]
Latest commit now has the core parameterisation returning `tendency` (radiative) and `delta`s (advective, autoconversion, sedimentation) for `t` and `q`, as well as the surface sedimentation.
`t`, `q`, and `prec_surf` are now updated in the interface routine.

Data transfer has also been reduced by passing only the part of the column that is operated on by the NN (`1:nrf`) - previously the entire column from the model was passed in).

Care had to be taken when passing `t` and `q` because Fortran is pass by reference and we do not want the semi-updated values returned - instead allocate and make a copy in the routine.

Routine operates on a generic grid as discussed above and in #42 
This involved changing the `y_in` input to be 2-D have a value for each atmospheric column (previously 1-D assuming a lat-lon grid).

**Comment by jatkinson1000** [comment]
Given CAM will have no interest in `q` or `t` (as they will be proxys for CAM variables) we could instead perform the allocation and copy in the SAM interface which would avoid wasted allocate and copy operations when running in CAM.

However, this is unlikely to be a bottleneck, and doing this would arguably violate the principle of least surprise should anyone come to re-apply the code somewhere that does deal in `t` and `q` and end up with some double counting.

**Comment by jatkinson1000** [comment]
@raehik Would you mind doing a review on this PR?

**Review by raehik — CHANGES_REQUESTED** [review]
Nice stuff. Couple questions, couple nits, main point on possibly missing final dimension extents when applying tendencies in SAM interface. LGTM otherwise!

File: NN_module/nn_convection_flux.f90 [hunk 1]
```diff
@@ -184,26 +187,14 @@ subroutine nn_convection_flux(tabs_i, q_i, y_in, &
...
 
         != unit K :: t
-        real, intent(inout) :: t(:, :, :)
+        real, intent(in) :: t_0(:, :, :)
             !! Liquid Ice static energy (cp*T + g*z − L(qliq + qice) − Lf*qice)
+        real, allocatable :: t(:, :, :)
         
         != unit 1 :: q
-        real, intent(inout) :: q(:, :, :)
+        real, intent(in) :: q_0(:, :, :)
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:195]: A fair point, have moved.
Also added some units, but this has not been fully completed/checked with CamFort. May be worth doing at a later point.

File: NN_module/nn_interface_CAM.f90 [hunk 2]
```diff
@@ -0,0 +1,146 @@
+module nn_interface_SAM_mod
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_interface_CAM.f90:1]: Good catch, have removed and will add into a later PR.

File: NN_module/nn_convection_flux.f90 [hunk 3]
```diff
@@ -217,16 +208,17 @@ subroutine nn_convection_flux(tabs_i, q_i, y_in, &
...
         ! NN outputs
         real,   dimension(nrf) :: t_flux_adv, q_flux_adv, q_tend_auto, &
-                                  q_sed_flux, t_rad_rest_tend
+                                  q_sed_flux
+        real, intent(out), dimension(:,:,:) :: t_rad_rest_tend
         ! Intermediate variables
-        real,   dimension(nrf) :: t_tendency_adv, q_tendency_adv, &
-                                  q_tendency_auto, q_tendency_sed, &
-                                  omp, fac
+        real, intent(out), dimension(:,:,:) :: t_delta_adv, q_delta_adv, &
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:217]: Moved as appropriate (with note left to say t_rad_rest_tend is also an output from the NN).

File: NN_module/nn_interface_SAM.f90 [hunk 4]
```diff
@@ -57,23 +61,33 @@ end subroutine nn_convection_flux_SAM_init
...
+        integer, intent(in) :: nx, ny, ny_gl, nstep, nstatis, icycle, YES3D
 
-        real :: y_in(ny)
+        real :: y_in(nx, ny)
             !! Distance of column from equator (proxy for insolation and sfc albedo)
 
+        real, dimension(nx, ny, nrf) :: t_rad_rest_tend, t_delta_adv, q_delta_adv, &
+                                        t_delta_auto, t_delta_sed, &
+                                        q_delta_auto, q_delta_sed
+        real, dimension(nx, ny)      :: prec_sed
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_interface_SAM.f90:88]: Done.

File: NN_module/nn_interface_SAM.f90 [hunk 5]
```diff
@@ -100,15 +114,50 @@ subroutine nn_convection_flux_SAM(tabs_i, q_i, &
...
+        ! advective, autoconversion (dt = -dq*(latent_heat/cp)),
+        ! sedimentation (dt = -dq*(latent_heat/cp)),
+        ! radiation rest tendency (multiply by dtn to get dt)
+        q(:,:,1:nrf) = q(:,:,1:nrf) + q_delta_adv(:,:,:) &
+                                    + q_delta_auto(:,:,:) &
+                                    + q_delta_sed(:,:,:)
+        t(:,:,1:nrf) = t(:,:,1:nrf) + t_delta_adv(:,:,:) &
+                                    + t_delta_auto(:,:,:) &
+                                    + t_delta_sed(:,:,:) &
+                                    + t_rad_rest_tend(:,:,:)*dtn
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_interface_SAM.f90:141]: So the previous version of the code passed the entire atmospheric column `(:,:,1:nz)` to the parameterisation, but only updated the lowest `nrf` cells of `t` and `q`.

As a minor improvement whilst refactoring I re-wrote such that the subroutine now takes inputs of size `(:,:,1:nrf)` and returns the same size, thereby reducing some unnecessary operations and arguably improving scientific clarity.

**Comment by jatkinson1000** [comment]
Thanks @raehik 
If you are happy with the changes/responses let me know and I'll merge to close.

**Comment by dorchard** [comment]
lgtm

**Pull request merged**